solvers/jaxopt_gd.py

```diff
@@ -0,0 +1,146 @@
+            if self.inner_solver == 'gd':
+                solver = jaxopt.GradientDescent(
+                    fun=f, maxiter=n_steps, implicit_diff=True,
+                    acceleration=False
+                )
+            elif self.inner_solver == 'lbfgs':
+                solver = jaxopt.LBFGS(
+                    fun=f, maxiter=n_steps, implicit_diff=True,
+                )
+            else:
+                raise ValueError(f"Inner solver {self.inner_solver} not"
+                                 + "available")
+            return solver.run(inner_var, outer_var).params
+
+        self.inner_solver_fun = partial(inner_solver_fun,
+                                        f=self.f_inner,
+                                        n_steps=self.n_inner_steps)
+        self.jaxopt_solver = partial(jaxopt_bilevel_solver,
+                                     inner_solver=self.inner_solver_fun)
+
+        self.inner_var0 = inner_var0
+        self.outer_var0 = outer_var0
+        self.run_once(2)
+
+    def run(self, callback):
+        eval_freq = self.eval_freq
+
+        # Init variables
+        inner_var = self.inner_var0.copy()
+        outer_var = self.outer_var0.copy()
+
+        step_sizes = jnp.array(
+            [self.step_size_outer]
+        )
+        exponents = jnp.zeros(1)
+        state_lr = init_lr_scheduler(step_sizes, exponents)
+
+        carry = dict(
+            state_lr=state_lr,
+        )
+
+        while callback((inner_var, outer_var)):
+            inner_var, outer_var, carry = self.jaxopt_solver(
+                    self.f_inner, self.f_outer, inner_var, outer_var,
+                    n_inner_steps=self.n_inner_steps, max_iter=eval_freq,
+                    **carry
+            )
+
+        self.beta = (inner_var, outer_var)
+
+    def get_result(self):
+        return self.beta
+
+
+@partial(jax.jit, static_argnums=(0, 1),
+         static_argnames=('n_inner_steps', 'max_iter', "inner_solver"))
+def jaxopt_bilevel_solver(f_inner, f_outer, inner_var, outer_var,
+                          state_lr=None, n_inner_steps=300,
+                          inner_solver=None,
+                          max_iter=1):
+    grad_outer = jax.grad(f_outer, argnums=(0, 1))
+
+    def jaxopt_one_iter(carry, _):
+        outer_lr, carry['state_lr'] = update_lr(
+            carry['state_lr']
+        )
+
+        carry['inner_var'], jvp_fun = jax.vjp(inner_solver,
+                                              carry['outer_var'],
+                                              carry['inner_var'])
+
+        grad_outer_in, grad_outer_out = grad_outer(carry['inner_var'],
+                                                   carry['outer_var'])
+
+        implicit_grad = grad_outer_out + jvp_fun(grad_outer_in)[0]
+        carry['outer_var'] -= outer_lr * implicit_grad
+
+        return carry, _
+
+    init = dict(
+        inner_var=inner_var, outer_var=outer_var, state_lr=state_lr
+    )
+    carry, _ = jax.lax.scan(
```

Comment:
As this is not a stochastic solver I am not sure it is necessary to have an jitted loop here.
I would simply put the for loop directly in the `run`

Reply:
Eventhough these are not stochastic solvers, jit-compilation seems much more efficient (in the case `jit == False`, the solver was run in the context manager `jax.disable_jit`. Maybe because we still have a large number of inner iterations in total (`n_outer_steps * n_inner_steps`)? What would be the gain of not using jitted-loops?
![image](https://github.com/benchopt/benchmark_bilevel/assets/77896657/7ee35c16-a638-4243-9ab7-2cf19d2aee49)

Reply:
Actually, even evaluation of a simple function is more efficient when the function is jitted 
![image](https://github.com/benchopt/benchmark_bilevel/assets/77896657/b086de91-922c-4130-8c45-a97cf6919c75)

Reply:
Is the above orange curve with jitted oracles + unjitted loop, or fully unjitted? 

I think that the oracles should always be jitted.

Reply:
Indeed with jitted oracles and unjitted loops the gain is insignificant
![image](https://github.com/benchopt/benchmark_bilevel/assets/77896657/2094a7e7-20af-458e-a44f-5687535e65b8)

Reply:
The gain of non-jitted code is that it is less complex (you can remove the carry, the scan and just write the loop in full in the solver).